File: FunikiSDKExample/FunikiSDKExample/FunikiSDKMotionViewController.m
//
//  Created by Matilde Inc.
//  Copyright (c) 2015 Fun'iki Project. All rights reserved.
//

#import "FunikiSDKMotionViewController.h"
#import "MAFunikiManager.h"
@interface FunikiSDKMotionViewController() <MAFunikiManagerDelegate, MAFunikiManagerDataDelegate>
{
    CGRect _baseSize;
    
    CGRect _accXRect;
    CGRect _accYRect;
    CGRect _accZRect;
    CGRect _rotXRect;
    CGRect _rotYRect;
    CGRect _rotZRect;
}

@property (weak, nonatomic) IBOutlet UIView *accXView;
@property (weak, nonatomic) IBOutlet UIView *accYView;
@property (weak, nonatomic) IBOutlet UIView *accZView;
@property (weak, nonatomic) IBOutlet UIView *rotXView;
@property (weak, nonatomic) IBOutlet UIView *rotYView;
@property (weak, nonatomic) IBOutlet UIView *rotZView;

@property (weak, nonatomic) IBOutlet UISwitch *sensorSwitch;

@end

@implementation FunikiSDKMotionViewController

#pragma mark - UIViewController
- (void)viewDidLoad
{
    [super viewDidLoad];
    
    _baseSize = self.accXView.frame;
    _baseSize.size.width = _baseSize.size.width / 2;
    
    _accXRect = self.accXView.frame;
    _accYRect = self.accYView.frame;
    _accZRect = self.accZView.frame;
    _rotXRect = self.rotXView.frame;
    _rotYRect = self.rotYView.frame;
    _rotZRect = self.rotZView.frame;
    
    [self setBarsHidden:YES];
}

- (void)viewWillAppear:(BOOL)animated
{
    [[MAFunikiManager sharedInstance] setDelegate:self];
    [[MAFunikiManager sharedInstance] setDataDelegate:self];
    
    [self updateSensorSwitch];
    
    [super viewWillAppear:animated];
}

- (BOOL)shouldAutorotate
{
    return NO;
}

-(void)setBarsHidden:(BOOL)hidden
{
    self.accXView.hidden = hidden;
    self.accYView.hidden = hidden;
    self.accZView.hidden = hidden;
    self.rotXView.hidden = hidden;
    self.rotYView.hidden = hidden;
    self.rotZView.hidden = hidden;
}

- (void)updateSensorSwitch
{
    if ([MAFunikiManager sharedInstance].isConnected){
        self.sensorSwitch.enabled = YES;
    }
    else {
        self.sensorSwitch.enabled = NO;
        [self.sensorSwitch setOn:NO animated:YES];
    }
}

#pragma mark - MAFunikiManagerDelegate
- (void)funikiManagerDidConnect:(MAFunikiManager *)manager
{
    NSLog(@"SDK Version%@", [MAFunikiManager funikiSDKVersionString]);
    NSLog(@"Firmware Revision%@", manager.firmwareRevision);
    
    [self updateSensorSwitch];
}

- (void)funikiManagerDidDisconnect:(MAFunikiManager *)manager error:(NSError *)error
{
    NSLog(@"%@", error);
    [self updateSensorSwitch];
}

- (void)funikiManager:(MAFunikiManager *)manager didUpdateCentralState:(CBCentralManagerState)state
{
    NSLog(@"CentralState:%ld", (long)state);
    
    [self updateSensorSwitch];
}

#pragma mark - MAFunikiManagerDelegate

- (void)funikiManager:(MAFunikiManager *)manager didUpdateMotionData:(MAFunikiMotionData *)motionData
{
    _accXRect.size.width = motionData.acceleration.x * _baseSize.size.width;
    _accYRect.size.width = motionData.acceleration.y * _baseSize.size.width;
    _accZRect.size.width = motionData.acceleration.z * _baseSize.size.width;
    _rotXRect.size.width = ((motionData.rotationRate.x * _baseSize.size.width) / 250.0) * 2.0;
    _rotYRect.size.width = ((motionData.rotationRate.y * _baseSize.size.width) / 250.0) * 2.0;
    _rotZRect.size.width = ((motionData.rotationRate.z * _baseSize.size.width) / 250.0) * 2.0;
    
    self.accXView.frame = _accXRect;
    self.accYView.frame = _accYRect;
    self.accZView.frame = _accZRect;
    self.rotXView.frame = _rotXRect;
    self.rotYView.frame = _rotYRect;
    self.rotZView.frame = _rotZRect;
}

- (void)funikiManager:(MAFunikiManager *)manager didPushButton:(MAFunikiManagerButtonEventType)buttonEventType
{
    NSString *message = @"";
    switch (buttonEventType) {
        case MAFunikiManagerButtonEventTypeSinglePush:
            message = @"ButtonEvnetTypeSinglePush";
            break;
            
        case MAFunikiManagerButtonEventTypeDoublePush:
            message = @"ButtonEventTypeDoublePush";
            break;
            
        default:
            message = @"ButtonEventTypeUnknown";
            break;
    }
    
    UIAlertView *alertView = [[UIAlertView alloc] initWithTitle:@"DidPushButton" message:message delegate:nil cancelButtonTitle:@"OK" otherButtonTitles:nil];
    [alertView show];
}

#pragma mark - Action
- (IBAction)switchDidChange:(UISwitch *)sender
{
    if (sender.on){
        // モーションセンサーを起動します。
        [[MAFunikiManager sharedInstance] startMotionSensor];
        [self setBarsHidden:NO];
    }else{
        // モーションセンサーを停止します。
        [[MAFunikiManager sharedInstance] stopMotionSensor];
        [self setBarsHidden:YES];
    }
}

@end

File: FunikiSDKExampleSwift/FunikiSDKExampleSwift/FunikiSDKMotionViewController.swift
//
//  Created by Matilde Inc.
//  Copyright (c) 2015 FUN'IKI Project. All rights reserved.
//

import UIKit

class FunikiSDKMotionViewController: UIViewController, MAFunikiManagerDelegate, MAFunikiManagerDataDelegate {
    
    let funikiManager = MAFunikiManager.sharedInstance()
    
    var baseSize:CGRect!
    var accXRect:CGRect!
    var accYRect:CGRect!
    var accZRect:CGRect!
    var rotXRect:CGRect!
    var rotYRect:CGRect!
    var rotZRect:CGRect!
    
    @IBOutlet var accXView:UIView!
    @IBOutlet var accYView:UIView!
    @IBOutlet var accZView:UIView!
    @IBOutlet var rotXView:UIView!
    @IBOutlet var rotYView:UIView!
    @IBOutlet var rotZView:UIView!
    
    @IBOutlet var sensorSwitch:UISwitch!

    // MARK: - UIViewController
    override func viewDidLoad() {
        baseSize = accXView.frame
        baseSize.size.width = baseSize.size.width / 2
        
        accXRect = accXView.frame
        accYRect = accYView.frame
        accZRect = accZView.frame
        
        rotXRect = rotXView.frame
        rotYRect = rotYView.frame
        rotZRect = rotZView.frame
        
        setBarHidden(hidden: true)
    }
        
    override func viewWillAppear(_ animated: Bool) {
        funikiManager?.delegate = self
        funikiManager?.dataDelegate = self
        
        self.updateSensorSwitch()
        
        super.viewWillAppear(animated)
    }
    
    override var shouldAutorotate: Bool{
        return false
    }

    
    // MARK: -
    func setBarHidden(hidden:Bool) {
        accXView.isHidden = hidden
        accYView.isHidden = hidden
        accZView.isHidden = hidden
        
        rotXView.isHidden = hidden
        rotYView.isHidden = hidden
        rotZView.isHidden = hidden
    }
    
    func updateSensorSwitch() {
        if (funikiManager?.isConnected)! {
            sensorSwitch.isEnabled = true
        }
        else {
            sensorSwitch.isEnabled = false
            sensorSwitch.setOn(false, animated: true)
        }
    }
    
    // MARK: - MAFunikiManagerDelegate
    func funikiManagerDidConnect(_ manager: MAFunikiManager!) {
        
        print("SDK Version\(MAFunikiManager.funikiSDKVersionString())")
        print("Firmware Revision\(manager.firmwareRevision)")
        
        updateSensorSwitch()
    }
    
    func funikiManagerDidDisconnect(_ manager: MAFunikiManager!, error: Error!) {
        if let actualError = error {
            print(actualError)
        }
        updateSensorSwitch()
    }
    
    func funikiManager(_ manager: MAFunikiManager!, didUpdateCentralState state: CBCentralManagerState) {
        updateSensorSwitch()
    }
    
    // MARK: - MAFunikiManagerDataDelegate
    func funikiManager(_ manager: MAFunikiManager!, didUpdate motionData: MAFunikiMotionData!) {
        
        accXRect.size.width = CGFloat(motionData.acceleration.x) * baseSize.size.width
        accYRect.size.width = CGFloat(motionData.acceleration.y) * baseSize.size.width
        accZRect.size.width = CGFloat(motionData.acceleration.z) * baseSize.size.width
        rotXRect.size.width = ((CGFloat(motionData.rotationRate.x) * baseSize.size.width) / 250.0) * 2.0
        rotYRect.size.width = ((CGFloat(motionData.rotationRate.y) * baseSize.size.width) / 250.0) * 2.0
        rotZRect.size.width = ((CGFloat(motionData.rotationRate.z) * baseSize.size.width) / 250.0) * 2.0
        
        self.accXView.frame = accXRect
        self.accYView.frame = accYRect
        self.accZView.frame = accZRect
        self.rotXView.frame = rotXRect
        self.rotYView.frame = rotYRect
        self.rotZView.frame = rotZRect
    }
    
    func funikiManager(_ manager: MAFunikiManager!, didPushButton buttonEventType: MAFunikiManagerButtonEventType) {
        
        var string = "ButtonEventType"
        switch (buttonEventType){
        case .singlePush:
            string = "ButtonEventTypeSinglePush"
            break
            
        case .doublePush:
            string = "ButtonEventTypeDoublePush"
            
        default:
            string = "ButtonEventTypeUnknown"
            break
        }
        
        let alertView = UIAlertView(title: "DidPushButton", message: string, delegate: nil, cancelButtonTitle: "OK")
        alertView.show()
    }
    
    // MARK: - Action
    @IBAction func switchDidChange(sender:UISwitch!) {
        
        if sensorSwitch.isOn {
            funikiManager?.startMotionSensor()
            setBarHidden(hidden: false)
        }
        else {
            funikiManager?.stopMotionSensor()
            setBarHidden(hidden: true)
        }
    }
}

